### PR TITLE
change to .phpcs.xml.dist and update rules

### PR DESCRIPTION
**Pull request by NicktheGeek**
The .phpcs.xml.dist rule file is detected and implemented automatically, reducing complexity for running phpcs.

The updated rules account for new wpcs sniff properties including the ability to check global prefix and i18n implementation with the correct text domain when specified.

The values have been left blank so the sniffs are generic, but the rules are in place to make it easier to get started by scaffold.

This is tangentially related to https://github.com/wp-cli/scaffold-command/issues/109

This is also loosely related to https://github.com/wp-cli/scaffold-command/issues/63 as updating composer scaffolds could work with this change.

@GaryJones may be able to provide additional feedback on this file change.

### CODE REVIEW

**Comment by schlessera** [comment]
Yes, I would indeed love feedback from @GaryJones and @jrfnl to know whether this would be a good default starting point for new plugins.

**Comment by schlessera** [comment]
@NicktheGeek For the blanks, it would be helpful to add comments so that developers immediately know where to turn the dials.

**Comment by NicktheGeek** [comment]
@schlessera good idea.

File: templates/.phpcs.xml.dist [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
+    <arg name="extensions" value="php"/>
+    <arg name="report" value="full"/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: Remove the reports args - the default is a full report anyway, and any different reports should be decided upon at runtime.

File: templates/.phpcs.xml.dist [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0"?>
+<ruleset name="WordPress Coding Standards for Plugins">
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: This naming is a bit misleading - perhaps `"WordPress Coding Standards for this plugin"? is better, and less implies that there's a WPCS specifically for Plugins?

File: templates/.phpcs.xml.dist [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
+    <arg name="extensions" value="php"/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: Another potentially useful arg is `<arg name="parallel" value="8"/>` - the number can vary, but 8 is a nice starting point.

File: templates/.phpcs.xml.dist [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    <description>Generally-applicable sniffs for WordPress plugins.</description>
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: Remove the space before the slash.

File: templates/.phpcs.xml.dist [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+<ruleset name="WordPress Coding Standards for Plugins">
+    <description>Generally-applicable sniffs for WordPress plugins.</description>
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: I know @jrfnl favours `./` here rather than `.`.

File: templates/.phpcs.xml.dist [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    <arg name="colors" />
+    <arg name="extensions" value="php"/>
+    <arg name="report" value="full"/>
+    <arg name="report" value="summary"/>
+    <arg name="report" value="source"/>
+
+	<!-- Rules: Check PHP version compatibility -->
+	<config name="testVersion" value="5.3-"/>
+	<rule ref="PHPCompatibility"/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: Since this is a WordPress plugin/theme, then it can reference the `PHPCompatibilityWP` ruleset instead - would need the `composer.json` template updated as well to pull that package in.

File: templates/.phpcs.xml.dist [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    <!-- Rules: WordPress Coding Standards -->
+    <config name="minimum_supported_wp_version" value="3.3"/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: This seems particularly low. I personally use current - 2 versions (so, 4.7 right now), and WPCS default is current - 3 versions (so, 4.6 for WPCS 1.0.0).

File: templates/.phpcs.xml.dist [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    <rule ref="WordPress">
+        <exclude name="WordPress.VIP"/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: Might be worth waiting until WPCS 1.0.0 drops, which already removes `WordPress-VIP` from the `WordPress` ruleset.

File: templates/.phpcs.xml.dist [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    </rule>
+    <rule ref="WordPress.NamingConventions.PrefixAllGlobals">
+        <properties>
+            <!-- Value: provide the function, class, and variable prefixes used. Separate multiple prefixes with a comma. -->
+            <property name="prefixes" type="array" value=""/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: @jrfnl may know better, but having a `my-plugin` placeholder here may better point folks towards changing it to something useful when it gives some warnings on the first run, rather than an empty value which may either blow things up, or not show any related violations (and therefore not drawing attention to changing it) at all.

File: templates/.phpcs.xml.dist [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        </properties>
+    </rule>
+    <rule ref="WordPress.WP.I18n">
+        <properties>
+            <!-- Value: provide the text domain used. -->
+            <property name="text_domain" type="array" value=""/>
```
Comment (GaryJones) [inline templates/.phpcs.xml.dist]: I think WPCS 1.0.0 now supports a better way of adding in array properties - individual entries, instead of a comma-separated value.

**Comment by GaryJones** [comment]
Rushing the review, and @jrfnl can fill in some blanks (or I can another day), but overall +1 for the PR to make it a much nicer starting point to implementing WPCS.

Thanks @NicktheGeek!

**Comment by NicktheGeek** [comment]
thanks you @GaryJones after you helped me get this on GFWA I've used it on several projects and I can't tell you the number of times it has found a minor typo in the text domain and saved me so much headache down the road. That one feature is worth the price of admission.

File: templates/.phpcs.xml.dist (hunk 2 above)
Comment (NicktheGeek) [reply templates/.phpcs.xml.dist]: I'm pulling this name from the current version, but I am good to do this naming change.

File: templates/.phpcs.xml.dist (hunk 6 above)
Comment (NicktheGeek) [reply templates/.phpcs.xml.dist]: Since that is outstanding, should this remain as is and implement a change for it with #63?

File: templates/.phpcs.xml.dist (hunk 1 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: FYI: I totally agree with @GaryJones' advice about the report.

File: templates/.phpcs.xml.dist (hunk 2 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: May I suggest "_WPCS based custom ruleset for plugins_" or "_WordPress Coding Standards based custom ruleset for your plugin/for plugin xyz_" ?

Makes it very clear that this is a _custom ruleset_ and not a _standard_. There is a significant difference between those two and naming this right will help people find the right information if they would search for more info.

To explain the difference:
* A **standard** can have add new sniffs, can be installed and referenced by name.
* A **custom ruleset** uses existing standards and ruleset and may add configuration for these.

File: templates/.phpcs.xml.dist (hunk 3 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: :+1: Comment to add along-side it: `<!-- Enabled parallel processing when available for faster results. -->`

File: templates/.phpcs.xml.dist (hunk 5 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: Either is fine as long as the `[.]phpcs.xml.dist` file is located in the project root.

File: templates/.phpcs.xml.dist (hunk 6 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: I would most definitely recommend changing this already to use the `PHPCompatibilityWP` standard instead as it will prevent false positives for things WP back-fills/poly-fills.

File: templates/.phpcs.xml.dist (hunk 8 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: This will work either way, even if the WPCS VIP deprecation would not make it into 1.0.0, so I'd leave it for now.

File: templates/.phpcs.xml.dist (hunk 9 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: > having a my-plugin placeholder here may better point folks towards changing it to something useful when it gives some warnings on the first run

I concur with @GaryJones, please do add a placeholder to give people incentive to change it.

> may either blow things up, or not show any related violations

An empty value won't blow things up, but will basically disable the sniff.

File: templates/.phpcs.xml.dist (hunk 10 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: Not WPCS, but PHPCS 3.3.0+. So if you would want to use the new array format, you would need to make sure that people are aware that they will need a very recent PHPCS version.

**Comment by NicktheGeek** [comment]
@jrfnl Thanks! I've added the changes you recommended.

@schlessera this may need some additional feedback from @jrfnl or @GaryJones but so far I've implemented almost everything suggested. There are a couple of comments regarding WPCS 1.0.0 but IMO we are better if this is implemented now and an issue created to update when 1.0.0 comes out with additional improvements. I'm good either way though pending release timeline for WP CLI 2.0 and WPCS 1.0.0.

File: templates/.phpcs.xml.dist [hunk 11]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0"?>
+<ruleset name="WordPress Coding Standards based custom ruleset for your plugin">
+    <description>Generally-applicable sniffs for WordPress plugins.</description>
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: Be aware that `exclude-pattern` values are interpreted as regular expressions, so this snippet would exclude *every* directory which included the string `vendor/`, i.e. `/some-other-vendor/` would also be excluded.

While this may be something which will rarely cause issues, you can stabilize this to just the Composer `vendor` directory by using something along the lines of `^/vendor/*` or the slightly more generic `./vendor/`

File: templates/.phpcs.xml.dist (hunk 10 above)
Comment (NicktheGeek) [reply templates/.phpcs.xml.dist]: @jrfnl will the current CSV method work with PHPCS 3.3.0+? If so it seams like it might be better to wait before implementing that even by reference.

File: templates/.phpcs.xml.dist [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+    <!-- How to scan -->
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: I would recommend adding a link here to the PHPCS wiki so people will know where to find more information about these options.

You can either use the generic wiki link: https://github.com/squizlabs/PHP_CodeSniffer/wiki

Or links to both these pages as good starting points for people:
* Usage instructions: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Usage
* Annotated ruleset: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml

File: templates/.phpcs.xml.dist [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="./"/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors"/>
+    <arg name="extensions" value="php"/>
+    <arg name="parallel" value="8"/><!-- Enables parallel processing when available for faster results. -->
+
+	<!-- Rules: Check PHP version compatibility -->
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: Indentation is off here. Tabs vs spaces.

Comment (jrfnl) [inline templates/.phpcs.xml.dist]: As per the earlier comments, using `PHPCompatibilityWP` will be the better choice.

I'd also recommend adding a link to the PHPCompatibilityWP repo as a comment, again to help people to more easily find out more info: https://github.com/PHPCompatibility/PHPCompatibilityWP

Maybe also add a link for more info about what the `testVersion` means and how it works: https://github.com/PHPCompatibility/PHPCompatibility#sniffing-your-code-for-compatibility-with-specific-php-versions

You may also want to add a comment saying that people can change the `testVersion` to suit their project.

File: templates/.phpcs.xml.dist [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	<config name="testVersion" value="5.3-"/>
+	<rule ref="PHPCompatibility"/>
+
+    <!-- Rules: WordPress Coding Standards -->
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: Along the same lines as some of my other comments, you may want to add a link here to the WPCS repo and/or to the WPCS wiki page about configuring sniffs:
* https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards
* https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties

File: templates/.phpcs.xml.dist [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+            <property name="prefixes" type="array" value="my-plugin"/>
+        </properties>
+    </rule>
+    <rule ref="WordPress.WP.I18n">
+        <properties>
+            <!-- Value: replace the text domain used. -->
+            <property name="text_domain" type="array" value="my-plugin"/>
+        </properties>
+    </rule>
+    <rule ref="WordPress.WhiteSpace.ControlStructureSpacing">
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: This is more a personal preference, than a recommended setting. I'm not sure if it belong in a scaffold.

File: templates/.phpcs.xml.dist [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        <properties>
+            <property name="blank_line_check" value="true"/>
+        </properties>
+    </rule>
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: I haven't looked at the scaffold in detail, but in case it also adds a PHPUnit test set-up, you may want to add an additional set of custom properties for that: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#custom-unit-test-classes

File: templates/.phpcs.xml.dist [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    <description>Generally-applicable sniffs for WordPress plugins.</description>
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="./"/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors"/>
```
Comment (jrfnl) [inline templates/.phpcs.xml.dist]: Just a side-note: `colors` will only work on certain platforms. Not sure by heart, but IIRC it works on Linux and Mac, but not on Windows.

**Comment by jrfnl** [comment]
Forgot to respond to this:

> There are a couple of comments regarding WPCS 1.0.0 but IMO we are better if this is implemented now and an issue created to update when 1.0.0 comes out with additional improvements.

WPCS 1.0.0 was due to come out last week, but there's still one issue which needs to be fixed. Expect it very soon though.

All the same, for the ruleset as it is, AFAICS, no changes would be needed for WPCS 1.0.0, so you're good.

File: templates/.phpcs.xml.dist (hunk 11 above)
Comment (jrfnl) [reply templates/.phpcs.xml.dist]: For more information about how the `exclude-pattern`s are interpreted, see: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-files-and-folders

Comment (jrfnl) [reply templates/.phpcs.xml.dist]: Darn... my bad. The more generic version should be `/vendor/`. The `.` which I'd inadvertently placed in front of it will not be interpreted as root, but as a regex wildcard, so, while it shouldn't give any problems, it also isn't necessary.

File: templates/.phpcs.xml.dist (hunk 1 above)
Comment (GaryJones) [reply templates/.phpcs.xml.dist]: Heh - the advice is what I took from Juliette!

**Comment by GaryJones** [comment]
The failure looks like some behat tests need updating to reference `.phpcs.xml.dist` (leading dot) instead of `phpcs.xml.dist`.

**Comment by NicktheGeek** [comment]
@GaryJones @jrfnl will it be ok for this to not have the leading .? I think it is required to be automatically used but I want to make sure I'm understanding the reason.

**Comment by jrfnl** [comment]
> @GaryJones @jrfnl will it be ok for this to not have the leading .? I think it is required to be automatically used but I want to make sure I'm understanding the reason.

Do you mean for the actual file name `phpcs.xml.dist` ? (in contrast to my remark about the exclude pattern https://github.com/wp-cli/scaffold-command/pull/161#discussion_r204142662 )

Yes, no problem.

Allowing a leading `.` for the filename was introduced in PHPCS 3.1.0 with a file loading order change in 3.1.1.

So for the custom ruleset with a leading dot to work, the minimum PHPCS requirement would have to be set at 3.1.0.

The file loading order (as of PHPCS 3.1.1) is:
1. `.phpcs.xml`
2. `phpcs.xml`
3. `.phpcs.xml.dist`
4. `phpcs.xml.dist`

In other words, not having the leading `.` now allows three different files to overrule the file.

Refs:
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.0
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.1

**Comment by GaryJones** [comment]
> will it be ok for this to not have the leading .?

And beyond what @jrfnl just said, whether it's right for PHPCS 3.1[.1] to be the minimum version for scaffolded plugins is a different matter, though for new plugins I don't see much reason why the default shouldn't be 3.3.

**Comment by NicktheGeek** [comment]
I went ahed and updated the tests to solve for the file ext change

File: templates/plugin-distignore.mustache [hunk 18]
```diff
@@ -21,7 +21,7 @@ phpunit.xml.dist
 multisite.xml
 multisite.xml.dist
 phpcs.xml
```
Comment (jrfnl) [inline templates/plugin-distignore.mustache]: You may also want to add `.phpcs.xml` here and possibly to the `.gitignore` file too ?

Comment (NicktheGeek) [reply templates/plugin-distignore.mustache]: @jrfnl thanks! I added .phpcs.xml but .gitignore is already in the file higher up.

**Comment by jrfnl** [comment]
FYI: in the mean time, both PHPCS 3.3.1 as well as WPCS 1.0.0 have been released.

**Comment by schlessera** [comment]
Thanks to @NicktheGeek for the pull request, and thanks to @GaryJones & @jrfnl for the extensive review!
